src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,936 @@
+//
+// For MLD specifically we can pull off some tricks to make evaluating alternatives fast:
+//   Only consider (s, c, t) with c border vertex: re-use MLD search steps.
+//   Add meta data to border vertices: consider (s, c, t) only when c is e.g. on a highway.
+//   Prune based on vertex cell id
+//
+// https://github.com/Project-OSRM/osrm-backend/issues/3905
+InternalManyRoutesResult alternativePathSearch(SearchEngineData<Algorithm> &search_engine_data,
+                                               const Facade &facade,
+                                               const PhantomNodes &phantom_node_pair,
+                                               unsigned number_of_alternatives)
+{
+    const auto max_number_of_alternatives =
+        std::min(facade.GetMaxNumberOfAlternatives(),               // profile-based upper bound
+                 static_cast<std::size_t>(number_of_alternatives)); // user-requested
+    const auto max_number_of_alternatives_to_unpack = facade.GetMaxNumberOfAlternativesToUnpack();
+    BOOST_ASSERT(max_number_of_alternatives > 0);
+    BOOST_ASSERT(max_number_of_alternatives_to_unpack >= max_number_of_alternatives);
+
+    const Partition &partition = facade.GetMultiLevelPartition();
+
+    // Prepare heaps for usage below. The searches will modify them in-place.
+    search_engine_data.InitializeOrClearFirstThreadLocalStorage(facade.GetNumberOfNodes());
+
+    Heap &forward_heap = *search_engine_data.forward_heap_1;
+    Heap &reverse_heap = *search_engine_data.reverse_heap_1;
+
+    // Do forward and backward search, save search space overlap as via candidates.
+    auto candidate_vias = makeCandidateVias(search_engine_data, facade, phantom_node_pair);
+
+    const auto by_weight = [](const auto &lhs, const auto &rhs) { return lhs.weight < rhs.weight; };
+    auto shortest_path_via_it =
+        std::min_element(begin(candidate_vias), end(candidate_vias), by_weight);
+
+    const auto has_shortest_path = shortest_path_via_it != end(candidate_vias) &&
+                                   shortest_path_via_it->node != SPECIAL_NODEID &&
+                                   shortest_path_via_it->weight != INVALID_EDGE_WEIGHT;
+
+    if (!has_shortest_path)
+        return InternalManyRoutesResult{};
+
+    NodeID shortest_path_via = shortest_path_via_it->node;
+    EdgeWeight shortest_path_weight = shortest_path_via_it->weight;
+
+    // Filters via candidate nodes with heuristics
+
+    // Note: filter pipeline below only makes range smaller; no need to erase items
+    // from the vector when we can mutate in-place and for filtering adjust iterators.
+    auto it = end(candidate_vias);
+
+    it = filterViaCandidatesByUniqueNodeIds(begin(candidate_vias), it);
+    it = filterViaCandidatesByRoadImportanceHeuristic(begin(candidate_vias), it, facade);
+    it = filterViaCandidatesByStretchHeuristic(begin(candidate_vias), it, shortest_path_weight);
```

Comment:
Nitpick: all of these are `heuristics`, so the it doesn't add information to the function name. Removing that could make the function name a little bit less unwieldy.